crates/sui/src/sui_commands.rs

```diff
@@ -638,41 +638,59 @@ async fn start(
+
     // If this is set, then no data will be persisted between runs, and a new genesis will be
     // generated each run.
-    if force_regenesis {
+    let config_dir = if force_regenesis {
         swarm_builder =
             swarm_builder.committee_size(NonZeroUsize::new(DEFAULT_NUMBER_OF_AUTHORITIES).unwrap());
         let genesis_config = GenesisConfig::custom_genesis(1, 100);
         swarm_builder = swarm_builder.with_genesis_config(genesis_config);
         let epoch_duration_ms = epoch_duration_ms.unwrap_or(DEFAULT_EPOCH_DURATION_MS);
         swarm_builder = swarm_builder.with_epoch_duration_ms(epoch_duration_ms);
+        tempdir()?.into_path()
     } else {
-        if config.is_none() && !sui_config_dir()?.join(SUI_NETWORK_CONFIG).exists() {
-            genesis(None, None, None, false, epoch_duration_ms, None, false).await.map_err(|_| anyhow!("Cannot run genesis with non-empty Sui config directory: {}.\n\nIf you are trying to run a local network without persisting the data (so a new genesis that is randomly generated and will not be saved once the network is shut down), use --force-regenesis flag.\nIf you are trying to persist the network data and start from a new genesis, use sui genesis --help to see how to generate a new genesis.", sui_config_dir().unwrap().display()))?;
-        }
-
-        // Load the config of the Sui authority.
-        // To keep compatibility with sui-test-validator where the user can pass a config
-        // directory, this checks if the config is a file or a directory
-        let network_config_path = if let Some(ref config) = config {
-            if config.is_dir() {
-                config.join(SUI_NETWORK_CONFIG)
-            } else if config.is_file()
-                && config
-                    .extension()
-                    .is_some_and(|ext| (ext == "yml" || ext == "yaml"))
+        // If the config path looks like a YAML file, it is treated as if it is the network.yaml
+        // overriding the network.yaml found in the sui config directry. Otherwise it is treated as
+        // the sui config directory for backwards compatibility with `sui-test-validator`.
+        let (network_config_path, sui_config_path) = match config {
+            Some(config)
+                if config.is_file()
+                    && config
+                        .extension()
+                        .is_some_and(|e| e == "yml" || e == "yaml") =>
             {
-                config.clone()
-            } else {
-                config.join(SUI_NETWORK_CONFIG)
+                (config, sui_config_dir()?)
+            }
+
+            Some(config) => (config.join(SUI_NETWORK_CONFIG), config),
+
+            None => {
+                let sui_config = sui_config_dir()?;
+                let network_config = sui_config.join(SUI_NETWORK_CONFIG);
+
+                if !network_config.exists() {
+                    genesis(None, None, None, false, epoch_duration_ms, None, false)
```

Comment:
Wouldn't this potentially override other settings in the `sui_config` dir?

Reply:
Ah, `genesis` will fail if the folder is not empty, so we're probably safe.

Reply:
This logic is pre-existing, but judging by the error message, genesis will not run if the directory is non-empty.

Reply:
Yes, I did that messy logic 😅  trying to preserve compatibility between `sui-test-validator` and `sui start`. :D